File: V8-1/JT-Loader_Vis-Studio_C#/JTLoader/Nodes/PropertyProxyMetaData.cs
/////////////////////////////////////////////////////////////////////
//
// This class contains the Property Proxy Meta Data used by the JT File
//
// Refer to Figure 132 page 178 - on (Ver 9.5 rev D)
//
/////////////////////////////////////////////////////////////////////


using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Drawing;
using System.Windows.Forms;

namespace C_sharp_JT_Reader
{
    public class PropertyProxyMetaData// : MetaDataNode
    {
        byte[] _MbString;
        Int32 _count;
        Int32 _elementLength;
        Int32 _filePosCount;
        byte[] _uncompressed;
        byte[] _fileBytes;
        float _fileVersion = 0;
        RichTextBox _richTextBox;
        
        List<string> m_textBox = new List<string>();

        System.Text.Encoding encoding = System.Text.Encoding.Unicode;

        public PropertyProxyMetaData(float fileVersion, RichTextBox richTextBox, int fileCount, byte[] uncompressed, int elementLength)
        {
            _fileVersion = fileVersion;
            _richTextBox = richTextBox;
            _filePosCount = fileCount;
            _uncompressed = uncompressed;
            _elementLength = elementLength;

            this.m_textBox.Add("\n\n-------------------------- Property Proxy Meta Data ------------------------------");
        }

        public int TraversePropertyProxyMetaData()
        {
            _fileBytes = new byte[4];
            // Loop until all of the meta data has been read & displayed
            while (_filePosCount < _elementLength)
            {
                Buffer.BlockCopy(_uncompressed, _filePosCount, _fileBytes, 0, 4);

                // count is part of the MbString type
                _count = DataTypes.getInt32(_fileBytes);

                _filePosCount += sizeof(Int32);

                _MbString = new byte[_count * 2];
                for (int i = 0; i < _MbString.Length; i++)
                {
                    _MbString[i] = _uncompressed[_filePosCount];
                    _filePosCount++;
                }
                string _propertyKey = encoding.GetString(_MbString);

                this.m_textBox.Add("\nProperty Key = " + _propertyKey);

                if (_propertyKey != null)
                {
                    byte _propertyValueType = _uncompressed[_filePosCount];
                    this.m_textBox.Add("\nProperty Value Type = " + _propertyValueType);
                    _filePosCount += sizeof(byte);
                    if (_propertyValueType == 1)
                    {
                        Buffer.BlockCopy(_uncompressed, _filePosCount, _fileBytes, 0, 4);

                        _count = DataTypes.getInt32(_fileBytes);

                        _filePosCount += sizeof(Int32);

                        _MbString = new byte[_count * 2];
                        for (int i = 0; i < _MbString.Length; i++)
                        {
                            _MbString[i] = _uncompressed[_filePosCount];
                            _filePosCount++;
                        }

                        string _propertyValue = encoding.GetString(_MbString);
                        this.m_textBox.Add("\nProperty Value = " + _propertyValue);
                    }
                    if (_propertyValueType == 2)
                    {
                        //I32 : Integer Property Value /////////////////////////////////////////////////////// To do /////////////////////////////////////////////////////////////////
                        _filePosCount += sizeof(Int32);
                    }
                    if (_propertyValueType == 3)
                    {
                        //F32 : Float Property Value   /////////////////////////////////////////////////////// To do /////////////////////////////////////////////////////////////////
                        //_filePosCount += sizeof(Int32);
                    }

                    // Date Property Value
                    if (_propertyValueType == 4)
                    {
                        // Copy 4 Bytes into destination array
                        Buffer.BlockCopy(_uncompressed, _filePosCount, _fileBytes, 0, 4);
                        Int16 Year = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);
                        Int16 Month = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);
                        Int16 Day = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);
                        Int16 Hour = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);
                        Int16 Minute = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);
                        Int16 Second = DataTypes.getInt16(_fileBytes);
                        _filePosCount += sizeof(Int16);

                        this.m_textBox.Add("\n" + Year.ToString() + Month.ToString() + Day.ToString() + Hour.ToString() + Minute.ToString() + Second.ToString());
                    }
                }

            }
            this.m_textBox.Add("\n");
            return this._filePosCount;
        }
    }
}

File: V8-1/JT-Loader_VSCode_TS/src/PropertyProxyMetaData.ts
import { Buffer } from "./BitBuffer";
import { DataTypes } from "./DataTypes";

export class PropertyProxyMetaData {
    _MbString: number[];
    _count: Int32;
    _elementLength: Int32;
    _filePosCount: Int32;
    _uncompressed: number[];
    _fileBytes: number[];
    _fileVersion: number = 0;
    m_textBox: List<string>;
    encoding: any;

    constructor(fileVersion: number, fileCount: number, uncompressed: number[], elementLength: number) {
        this._fileVersion = fileVersion;
        this._filePosCount = fileCount;
        this._uncompressed = uncompressed;
        this._elementLength = elementLength;
    }

    public TraversePropertyProxyMetaData(): number {
        this._fileBytes = new Array(4);
        while (this._filePosCount < this._elementLength) {
            Buffer.BlockCopy(this._uncompressed, this._filePosCount, this._fileBytes, 0, 4);
            this._count = DataTypes.getInt32(this._fileBytes);
            this._filePosCount += 4;
            this._MbString = new Array(this._count * 2);
            for (var i: number = 0; i < this._MbString.length; i++) {
                this._MbString[i] = this._uncompressed[this._filePosCount];
                this._filePosCount++;
            }
            var _propertyKey: string = this.encoding.GetString(this._MbString);
            if (_propertyKey != null) {
                var _propertyValueType: number = this._uncompressed[this._filePosCount];
                this._filePosCount += 1;
                if (_propertyValueType == 1) {
                    Buffer.BlockCopy(this._uncompressed, this._filePosCount, this._fileBytes, 0, 4);
                    this._count = DataTypes.getInt32(this._fileBytes);
                    this._filePosCount += 4;
                    this._MbString = new Array(this._count * 2);
                    for (var i: number = 0; i < this._MbString.length; i++) {
                        this._MbString[i] = this._uncompressed[this._filePosCount];
                        this._filePosCount++;
                    }
                    var _propertyValue: string = this.encoding.GetString(this._MbString);
                }
                if (_propertyValueType == 2) {
                    this._filePosCount += 4;
                }
                if (_propertyValueType == 3) {

                }
                if (_propertyValueType == 4) {
                    Buffer.BlockCopy(this._uncompressed, this._filePosCount, this._fileBytes, 0, 4);
                    var Year: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                    var Month: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                    var Day: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                    var Hour: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                    var Minute: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                    var Second: Int16 = DataTypes.getInt16(this._fileBytes);
                    this._filePosCount += 2;
                }
            }
        }
        return this._filePosCount;
    }
}